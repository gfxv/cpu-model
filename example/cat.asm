org 5

_start:

    loop:
        ld 1     ; read next char
        st #1    ; write to output buffer
        jmp loop

