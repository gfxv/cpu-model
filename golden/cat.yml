source: |-
  org 5
  counter:
      int 0

  _start:

      ld 1
      st &counter

      loop:
          ld 1
          st #1

          ld counter
          dec
          st &counter

          jz stop
          jmp loop

      stop:
          hlt
in: |-
  9hmmm.....
out_code: |-
  [
    {
      "index": 0,
      "opcode": "jmp",
      "arg": "7",
      "arg_type": "default"
    },
    {
      "index": 6,
      "opcode": "int",
      "arg": 0,
      "arg_type": "int"
    },
    {
      "index": 7,
      "opcode": "ld",
      "arg": "1",
      "arg_type": "default"
    },
    {
      "index": 8,
      "opcode": "st",
      "arg": 6,
      "arg_type": "raw"
    },
    {
      "index": 9,
      "opcode": "ld",
      "arg": "1",
      "arg_type": "default"
    },
    {
      "index": 10,
      "opcode": "st",
      "arg": "1",
      "arg_type": "raw"
    },
    {
      "index": 11,
      "opcode": "ld",
      "arg": 6,
      "arg_type": "default"
    },
    {
      "index": 12,
      "opcode": "dec",
      "arg_type": "none",
      "arg": 0
    },
    {
      "index": 13,
      "opcode": "st",
      "arg": 6,
      "arg_type": "raw"
    },
    {
      "index": 14,
      "opcode": "jz",
      "arg": 16,
      "arg_type": "default"
    },
    {
      "index": 15,
      "opcode": "jmp",
      "arg": 9,
      "arg_type": "default"
    },
    {
      "index": 16,
      "opcode": "hlt",
      "arg_type": "none",
      "arg": 0
    }
  ]
out: |
  Instructions: 66
  Ticks: 359
  === OUT ===
  --- bytes ---
  104 109 109 109 46 46 46 46 46
  --- string ---
  hmmm.....
log: |
  DEBUG    root:control_unit.py:109 TICK: 4    | jmp  | ARG: 7   | TYPE: default | PC: 7   | ACC: 0   | AR: 0   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 10   | ld   | ARG: 1   | TYPE: default | PC: 8   | ACC: 9   | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 17   | st   | ARG: 6   | TYPE: raw     | PC: 9   | ACC: 9   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 23   | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 104 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 30   | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 104 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 36   | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 9   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 40   | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 8   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 47   | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 8   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 51   | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 8   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 55   | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 8   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 61   | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 68   | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 74   | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 8   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 78   | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 7   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 85   | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 7   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 89   | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 7   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 93   | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 7   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 99   | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 106  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 112  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 7   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 116  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 6   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 123  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 6   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 127  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 6   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 131  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 6   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 137  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 144  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 109 | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 150  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 6   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 154  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 5   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 161  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 5   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 165  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 5   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 169  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 5   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 175  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 182  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 188  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 5   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 192  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 4   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 199  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 4   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 203  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 4   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 207  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 4   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 213  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 220  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 226  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 4   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 230  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 3   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 237  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 3   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 241  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 3   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 245  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 3   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 251  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 258  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 264  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 3   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 268  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 2   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 275  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 2   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 279  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 2   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 283  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 2   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 289  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 296  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 302  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 2   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 306  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 1   | AR: 12  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 313  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 1   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 317  | jz   | ARG: 16  | TYPE: default | PC: 15  | ACC: 1   | AR: 14  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 321  | jmp  | ARG: 9   | TYPE: default | PC: 9   | ACC: 1   | AR: 15  | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 327  | ld   | ARG: 1   | TYPE: default | PC: 10  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 334  | st   | ARG: 1   | TYPE: raw     | PC: 11  | ACC: 46  | AR: 1   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 340  | ld   | ARG: 6   | TYPE: default | PC: 12  | ACC: 1   | AR: 6   | N: False  | Z: False 
  DEBUG    root:control_unit.py:109 TICK: 344  | dec  | ARG: -   | TYPE: none    | PC: 13  | ACC: 0   | AR: 12  | N: False  | Z: True  
  DEBUG    root:control_unit.py:109 TICK: 351  | st   | ARG: 6   | TYPE: raw     | PC: 14  | ACC: 0   | AR: 6   | N: False  | Z: True  
  DEBUG    root:control_unit.py:109 TICK: 355  | jz   | ARG: 16  | TYPE: default | PC: 16  | ACC: 0   | AR: 14  | N: False  | Z: True  
  DEBUG    root:control_unit.py:109 TICK: 359  | hlt  | ARG: -   | TYPE: none    | PC: 17  | ACC: 0   | AR: 16  | N: False  | Z: True  
